FIO vb and secure boot for imx7ulpea

ldts opened the pull request:
This PR replaces AVB with FIOVB TA.

ldts committed:
[FIO toup] mmc: fsl_esdhc_imx: initialize data for imx7ulp

changes done as per the linux kernel

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] drivers: optee: rpmb: fix returning CID to TEE

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO internal] common: foundries.io verified boot utility

ricardosalveti commented on drivers/mmc/rpmb.c:
```diff
@@ -223,6 +223,7 @@ static void rpmb_hmac(unsigned char *key, unsigned char *buff, int len,
 	/* finish up 2nd pass */
 	sha256_finish(&ctx, output);
 }
+
```
No need for this extra line (makes rebasing harder later).

ricardosalveti commented on drivers/mmc/rpmb.c:
```diff
@@ -380,52 +385,68 @@ static int rpmb_route_write_req(struct mmc *mmc, struct s_rpmb *req,
 	/*
 	 * Read the result of the request.
 	 */
-
```
Same here.

ricardosalveti commented on drivers/mmc/rpmb.c:
```diff
@@ -439,28 +460,37 @@ static int rpmb_route_frames(struct mmc *mmc, struct s_rpmb *req,
-
+		ret = rpmb_route_write_req(mmc, req, req_cnt, rsp, rsp_cnt);
+		if (ret)
+			debug("error: RPMB_REQ_KEY\n");
```
Should this be a debug or an error message? Same for the other checks.

ricardosalveti commented on cmd/Kconfig:
```diff
@@ -2078,6 +2078,19 @@ config CMD_AVB
+	    fiovb init - initialize avb2 subsystem
+	    fiovb read_rb - read rollback index
+	    fiovb write_rb - write rollback index
+	    fiovb is_unlocked - check device lock state
```
As I added to the optee pull request, no need for the lock state here.

ricardosalveti commented on cmd/Makefile:
```diff
@@ -173,6 +173,9 @@ obj-$(CONFIG_CMD_BLOB) += blob.o
 # Android Verified Boot 2.0
 obj-$(CONFIG_CMD_AVB) += avb.o
 
+# FIO Verified Boot 2.0
```
2.0? :-)

Probably good to replace FIO to Foundries.IO here as well.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+
```
No need for this extra empty line.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Add 2019, Foundries.IO.

ldts replied:
add another copyright line?

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+#include <mmc.h>
+#include <fiovb.h>
+
+static struct FIOvbOps *fiovb_ops;
```
Is camel case really supported by u-boot?

ldts replied:
nah I dont think so. looks horrible to me but that is what was merged so I thought it was better just to follow this as a guideline. anyway, will change it.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	if (fiovb_ops)
+		return CMD_RET_SUCCESS;
+
+	printf("Failed to initialize fiovb1\n");
```
1 is the default, just change to 'Failed to initialize fiovb'.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	u64 rb_idx;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'fiovb init' first\n");
```
Remove 1.0 here, not needed.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	u64 rb_idx;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'fiovb init' first\n");
```
Same here.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	char *endp;
+
+	if (!fiovb_ops) {
+		printf("FIOVB 1.0 is not initialized, run 'avb init' first\n");
```
Here as well. Message is also saying avb instead of fiovb.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	const char *value;
+
+	if (!fiovb_ops) {
+		printf("AVB 2.0 is not initialized, run 'avb init' first\n");
```
s/avb/fio/g and drop 2.0.

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+
+U_BOOT_CMD(
+	fiovb, 29, 0, do_fiovb,
+	"Provides commands for testing Android Verified Boot 2.0 functionality",
```
"for testing Foundries.IO verified boot".

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+U_BOOT_CMD(
+	fiovb, 29, 0, do_fiovb,
+	"Provides commands for testing Android Verified Boot 2.0 functionality",
+	"init <dev> - initialize avb2 for <dev>\n"
```
s/avb2/fiovb/

ricardosalveti commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,221 @@
+	"avb write_rb <num> <rb> - write rollback index <rb> to <num>\n"
+	"avb is_unlocked - returns unlock status of the device\n"
+	"avb read_pvalue <name> <bytes> - read a persistent value <name>\n"
+	"avb write_pvalue <name> <value> - write a persistent value <name>\n"
```
Helper all saying 'avb' instead of fiovb.

ricardosalveti commented on common/fiovb.c:
```diff
@@ -0,0 +1,329 @@
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Add our own copyright.

ricardosalveti commented on include/fiovb.h:
```diff
@@ -0,0 +1,153 @@
+
```
Remove the first empty line.

ricardosalveti commented on include/fiovb.h:
```diff
@@ -0,0 +1,153 @@
+
+/*
+ * (C) Copyright 2018, Linaro Limited
```
Add our own copyright.

ricardosalveti commented on include/tee/optee_ta_fiovb.h:
```diff
@@ -0,0 +1,65 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2018, Linaro Limited */
```
Missing our own copyright.

ricardosalveti commented on configs/mx7ulpea-ucom_lmp_defconfig:
```diff
@@ -37,10 +38,12 @@ CONFIG_CMD_MEMTEST=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+# CONFIG_CMD_MMC_RPMB is not set
```
Do we need to disable this?

ldts replied:
no we dont need to. ok

ricardosalveti commented on configs/mx7ulpea-ucom_lmp_defconfig:
```diff
@@ -73,7 +77,9 @@ CONFIG_FSL_LPUART=y
 CONFIG_FSL_QSPI=y
+CONFIG_TEE=y
 CONFIG_OPTEE=y
+# CONFIG_OPTEE_TA_AVB is not set
```
Same here, is this on by default?

ldts replied:
seems to be set when I select some other option

ricardosalveti commented on common/Kconfig:
```diff
@@ -937,7 +937,7 @@ config HASH
 
 config AVB_VERIFY
 	bool "Build Android Verified Boot operations"
-	depends on LIBAVB && FASTBOOT
```
We can drop this patch as we will not be using avb anymore.

ldts replied:
sure. I thought we needed avb for the provisioning of the keys (although now that I remember we are only going to be provisioning from linux) ok

ricardosalveti commented on common/avb_verify.c:
```diff
@@ -356,7 +356,7 @@ static struct mmc_part *get_partition(AvbOps *ops, const char *partition)
 		return NULL;
 
-	dev_num = get_boot_device(ops);
+	dev_num = avb_get_boot_device(ops);
```
Also another patch we can drop if we are not using.

ricardosalveti commented on configs/mx7ulpea-ucom_lmp_defconfig:
```diff
@@ -8,6 +8,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_SPL_STACK_R_ADDR=0x68800000
 CONFIG_ARMV7_BOOT_SEC_DEFAULT=y
 CONFIG_SPL_TEXT_BASE=0x2F011000
+CONFIG_ANDROID_BOOT_IMAGE=y
```
Also fine to drop as we're not going to use it.

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -95,7 +298,6 @@ int board_init(void)
 #ifdef CONFIG_FSL_QSPI
 	board_qspi_init();
 #endif
-
```
Do we need this line removed?

ricardosalveti replied:
Had the same comment, @ldts mind removing the removal?

mike-scott commented on board/embeddedartists/mx7ulpea-ucom/mx7ulpea-ucom.c:
```diff
@@ -13,11 +13,25 @@
+#include <fs.h>
 #include <usb.h>
 #include <dm.h>
+#include <spi_flash.h>
```
Should we have these protected with? `#ifdef CONFIG_FPGA` -- I know we're enabling the CONFIG in `include/configs/mx7ulpea-ucom.h` but we do an `#ifdef` check below in the source.

ldts replied:
we can follow that path (@ricardosalveti  did that with the SPL include) but IMO it makes the code a bit uglier and for not much gain...anyhow let me know if you think you'd better see that for consistency

ricardosalveti replied:
Yeah, seems the standard in u-boot is to just include the header when it is building for SPL, so probably better to protect here as well.

mike-scott commented on drivers/tee/optee/rpmb.c:
```diff
@@ -106,6 +108,10 @@ static u32 rpmb_get_dev_info(u16 dev_id, struct rpmb_dev_info *info)
 
 	memcpy(info->cid, mmc->cid, sizeof(info->cid));
+
+	for (i = 0; i < 4; i++)
```
I feel like the `4` here should be the array length of info->cid (# of u32 elements)

ldts replied:
yep, but array length of mmc->cid instead

ldts replied:
I'll write it differently and post it upstream

mike-scott commented on drivers/mmc/rpmb.c:
```diff
@@ -380,52 +384,69 @@ static int rpmb_route_write_req(struct mmc *mmc, struct s_rpmb *req,
 	/*
 	 * Send the read request.
 	 */
+
```
Do we need this extra line?

mike-scott reviewed (changes requested):
Also:
commit: `[FIO toup] rpmb: use cache aligned buffers on route commands` needs a better description and probably split into 2-3 commits:
- handle aligned buffer access
- there are quite a few MMC cmd request changes like `MMC_RSP_R1` -> `MMC_RSP_R1b` which also need explanation

ldts commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,193 @@
+	if (argc != 3)
+		return CMD_RET_USAGE;
+
+	index = (size_t)simple_strtoul(argv[1], NULL, 16);
```
needs fixing to simple_strtoull

ldts commented on cmd/fiovb.c:
```diff
@@ -0,0 +1,193 @@
+		return CMD_RET_USAGE;
+
+	index = (size_t)simple_strtoul(argv[1], NULL, 16);
+	rb_idx = simple_strtoul(argv[2], NULL, 16);
```
bug. needs strtoull

ricardosalveti commented on configs/mx7ulpea-ucom_lmp_defconfig:
```diff
@@ -21,6 +21,7 @@ CONFIG_BOOTCOMMAND="fatload mmc 0:1 ${loadaddr} /boot.itb; setenv verify 1; sour
 CONFIG_DEFAULT_FDT_FILE="imx7ulpea-ucom-kit_v2"
 CONFIG_BOUNCE_BUFFER=y
 CONFIG_BOARD_EARLY_INIT_F=y
+CONFIG_FIOVB=y
```
Please also add the same changes to configs/mx7ulpea-ucom-mfgtool_defconfig, as that is the one used by the mfgtool build.

ldts committed:
[FIO toup] drivers: rpmb: replicate linux mmc configuration

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] drivers: rpmb: use cache aligned buffers on route commands

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO ext] configs: mx7ulpea-ucom: enable IMX_FPGA

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO toup] fpga: imx_m4: boot M4

This driver allows a controlled boot of the M4 processor.
When secure mode is enabled (FIOVB), the hash of the firmware in the
FIT image is compared against the value stored in the mmc RPMB
partition (via OPTEE). If they differ, the firmware in QSPI flash will
be upgraded to the new image before booting.

When secure mode is not enabled (!FIOVB), the hash of the firmware in
the FIT image is compared against the value calculated from QSPI flash
(but using the size of the FIT image firmware as the size of the QSPI
image in flash). If they difer, the firmware in QSPI flash will be
upgraded to the new image before booting.

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>

ldts committed:
[FIO ext] configs: mx7ulpea-mfgtool-ucom: enable IMX_FPGA

mike-scott reviewed (approved):
LGTM
